File: apps/loomx-api/src/services/sqlTableExtractor.ts
/**
 * Extracts table names referenced in a SQL query.
 *
 * Parses every FROM and JOIN clause to collect schema.table references.
 * Used as a server-side fallback when the frontend does not supply
 * tables_used — guarantees the field is never null in query_history.
 *
 * Returns a sorted, deduplicated array of normalised "schema.table" strings
 * (brackets and excess whitespace removed, lowercased).
 */
export function extractTablesFromSql(sql: string): string[] {
  const tables = new Set<string>();

  // Strip single-line and multi-line SQL comments then collapse whitespace.
  const stripped = sql
    .replace(/--[^\n]*/g, ' ')
    .replace(/\/\*[\s\S]*?\*\//g, ' ')
    .replace(/\s+/g, ' ')
    .trim();

  // Match FROM or any JOIN variant followed by a table reference.
  // Deliberately excludes subqueries that start with "(" so that
  // "FROM (SELECT ...)" does not produce a spurious match.
  //
  // Captures:
  //   [schema].[table]   →  schema.table
  //   dbo.MyTable        →  dbo.mytable
  //   MyTable            →  mytable
  //   [MyTable]          →  mytable
  const pattern =
    /\b(?:FROM|(?:INNER\s+|LEFT\s+|RIGHT\s+|FULL\s+|CROSS\s+|OUTER\s+)?JOIN)\s+(?!\()(\[?[\w$]+\]?(?:\s*\.\s*\[?[\w$]+\]?)?)/gi;

  let match: RegExpExecArray | null;
  while ((match = pattern.exec(stripped)) !== null) {
    const raw = match[1]
      .replace(/\[|\]/g, '') // strip square brackets
      .replace(/\s+/g, '')   // remove any whitespace around the dot
      .trim()
      .toLowerCase();

    // Skip SQL keywords that can appear after FROM/JOIN in edge cases,
    // temp tables (#tableName), variables (@var), and empty strings.
    if (
      raw &&
      raw.length >= 1 &&
      !raw.startsWith('#') &&
      !raw.startsWith('@') &&
      !/^(select|where|on|and|or|set|values|openquery|opendatasource|openrowset)$/.test(raw)
    ) {
      tables.add(raw);
    }
  }

  return Array.from(tables).sort();
}

File: apps/kaveon-api/services/sql_table_extractor.py
"""
Extract table names referenced in a SQL query.
Port of sqlTableExtractor.ts — identical regex logic.
"""

import re
from typing import List


def extract_tables_from_sql(sql: str) -> List[str]:
    """
    Parse every FROM and JOIN clause to collect schema.table references.
    Returns a sorted, deduplicated list of normalised "schema.table" strings.
    """
    tables = set()

    # Strip single-line and multi-line SQL comments, then collapse whitespace
    stripped = re.sub(r"--[^\n]*", " ", sql)
    stripped = re.sub(r"/\*[\s\S]*?\*/", " ", stripped)
    stripped = re.sub(r"\s+", " ", stripped).strip()

    pattern = re.compile(
        r"\b(?:FROM|(?:INNER\s+|LEFT\s+|RIGHT\s+|FULL\s+|CROSS\s+|OUTER\s+)?JOIN)"
        r"\s+(?!\()(\[?[\w$]+\]?(?:\s*\.\s*\[?[\w$]+\]?)?)",
        re.IGNORECASE,
    )

    _SKIP = {"select", "where", "on", "and", "or", "set", "values",
             "openquery", "opendatasource", "openrowset"}

    for match in pattern.finditer(stripped):
        raw = match.group(1)
        raw = re.sub(r"[\[\]]", "", raw)   # strip brackets
        raw = re.sub(r"\s+", "", raw)       # remove whitespace around dot
        raw = raw.strip().lower()

        if (
            raw
            and len(raw) >= 1
            and not raw.startswith("#")
            and not raw.startswith("@")
            and raw not in _SKIP
        ):
            tables.add(raw)

    return sorted(tables)
